### PR TITLE
Functors task 2 - OSennikov

### PR DESCRIPTION
**Pull request by JohnDou1991**

### CODE REVIEW

File: functors task (2)/student/include/impl.h [hunk 1]
```diff
@@ -21,7 +21,7 @@ class DataBrowser
      */
     void userEnter(const std::string& userId)
     {  // awful, but in header for testing
-        m_dataReaders.emplace(userId, getSelector());
+        m_dataReaders.emplace( userId, std::unique_ptr<IDataSelector>( getSelector() ) );
```
Comment (smay1613) [inline functors task (2)/student/include/impl.h:24]: cheating, please don't modify this code

or you mean something is wrong in task interface?

File: functors task (2)/student/src/impl.cpp [hunk 2]
```diff
@@ -7,48 +7,47 @@ void DataBrowser::userLeave(const std::string &userId)
...
-    }
-    if (it->second == nullptr)
-    {
-        return false;
-    }
-    return it->second->getDataType1(returnValues, 0);
+        return selector->getDataType1( returnValues, 0 );
+    };
+
+    return safeCall( userId, [this, &returnValues, function]( const std::unique_ptr<IDataSelector>& selector ) -> bool
```
Comment (smay1613) [inline functors task (2)/student/src/impl.cpp:15]: Hard to read. Consider naming this lambda.
```suggestion
    auto getter = [this, &returnValues, function]( const std::unique_ptr<IDataSelector>& selector ) ...
    return safeCall( userId, getter)
```

File: functors task (2)/student/src/impl.cpp [hunk 3]
```diff
@@ -7,48 +7,47 @@ void DataBrowser::userLeave(const std::string &userId)
...
-    if (it->second == nullptr)
+    auto function = []( IDataSelector* selector, std::vector<size_t>& returnValues ) -> bool
     {
-        return false;
-    }
-    return it->second->getDataType2(returnValues);
+        return selector->getDataType2( returnValues );
+    };
+
+    return safeCall( userId, [this, &returnValues, function]( const std::unique_ptr<IDataSelector>& selector ) -> bool
```
Comment (smay1613) [inline functors task (2)/student/src/impl.cpp:29]: here too

File: functors task (2)/student/include/impl.h (hunk 1 above)
Comment (JohnDou1991) [reply functors task (2)/student/include/impl.h:24]: It was not compiling otherwise.

Comment (smay1613) [reply functors task (2)/student/include/impl.h:24]: Please provide compilation log and compiler version

**Pull request closed without merging**